### PR TITLE
[Reward Center] An Auctioneer Program That Rewards Buyers and Sellers For NFT Sales

### PR DESCRIPTION
**Pull request by kespinola**
### Overview

Reward Center is an auctioneer program that payouts spl token from a treasury managed by the reward center to buyers and sellers for a sale.

### Changes
- Rust SDK for auction-house that is used to setup tests for reward center program
- Rust SDK for reward-center used to form instructions for testing the program using program test
- Test specs covering the "happy path" of every supported instruction
- More details on the program available in its README

### CODE REVIEW

File: listing-rewards/program/src/state.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    pub reward_center: Pubkey,
+    pub seller: Pubkey,
+    pub metadata: Pubkey,
+    pub price: u64,
+    pub token_size: u64,
+    pub bump: u8,
+    pub created_at: i64,
+    pub canceled_at: Option<i64>,
+    pub purchased_at: Option<i64>,
+    pub reward_redeemed_at: Option<i64>,
```
Comment (kespinola) [inline listing-rewards/program/src/state.rs:73]:
```suggestion
```

drop this field. no longer needed.

File: listing-rewards/program/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        mpl_token_metadata::ID
+    }
+}
+
+impl Deref for MetadataAccount {
+    type Target = mpl_token_metadata::state::Metadata;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
```
[the remark this answers is not in the document]
Comment (sunguru98) [reply listing-rewards/program/src/lib.rs]: Done.

File: listing-rewards/program/src/listings/cancel/mod.rs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            })
+            .collect(),
+    };
+
+    invoke_signed(
+        &cancel_listing_ix,
+        &cancel_listing_ctx_accounts.to_account_infos(),
+        reward_center_signer_seeds,
+    )?;
+
```
[the remark this answers is not in the document]
Comment (sunguru98) [reply listing-rewards/program/src/listings/cancel/mod.rs]: Done.

File: listing-rewards/program/src/state.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+#[account]
+#[derive(Debug)]
+pub struct RewardCenter {
+    /// the mint of the token used as rewards
+    pub token_mint: Pubkey,
+    /// the auction house associated to the reward center
+    pub auction_house: Pubkey,
+    /// the oracle allowed to adjust rewardable collections
+    pub collection_oracle: Option<Pubkey>,
```
[the remark this answers is not in the document]
Comment (sunguru98) [reply listing-rewards/program/src/state.rs]: Done.

File: listing-rewards/program/src/state.rs (hunk 1 above)
Comment (sunguru98) [reply listing-rewards/program/src/state.rs]: Done.

File: auction-house/program/src/withdraw/mod.rs [hunk 5]
```diff
@@ -117,7 +117,7 @@ pub fn withdraw<'info>(
 pub struct AuctioneerWithdraw<'info> {
     /// CHECK: Validated in withdraw_logic.
     /// User wallet account.
-    pub wallet: UncheckedAccount<'info>,
+    pub wallet: Signer<'info>,
```
[the remark this answers is not in the document]
Comment (sunguru98) [reply auction-house/program/src/withdraw/mod.rs]: Done.

**Comment by kespinola** [comment]
@austbot we are moving reward center over to holaplex for the time being. Once we've ran it in production for sometime we'll open a proposal for the metaplex community to vote on adopting it into mpl.

**Pull request closed without merging**